master/buildbot/worker/gce.py

```diff
@@ -0,0 +1,228 @@
+
+    def getMetadataFromState(self, state):
+        metadata = state['metadata']
+        dict = {}
+        for x in metadata.get('items', []):
+            dict[x['key']] = x['value']
+        return (metadata['fingerprint'], dict)
+
+    def getDesiredMetadata(self, build):
+        if ":" in self.masterFQDN:
+            host, port = self.masterFQDN.split(":")
+        else:
+            host = self.masterFQDN
+            port = 9989
+
+        return {
+            "WORKERNAME": self.name,
+            "WORKERPASS": self.password,
+            "BUILDMASTER": host,
+            "BUILDMASTER_PORT": port
+        }
+
+    def updateMetadata(self, build, metadata):
+        result = metadata.copy()
+        for key in GCE_NODE_METADATA_KEYS:
+            result.pop(key, None)
+        result.update(self.getDesiredMetadata(build))
+        return result
+
+    def getMetadataItemsFromDict(self, metadata):
+        return [{"key": key, "value": metadata[key]} for key in metadata]
+
+    def setMetadata(self, fingerprint, metadata):
+        return self._gce.setMetadata(fingerprint=fingerprint,
+            items=self.getMetadataItemsFromDict(metadata))
+
+    def getCurrentDiskName(self, disks):
+        for d in disks:
+            if d['boot']:
+                return d['deviceName']
+
+    def getNewDiskName(self, metadata):
+        gen = int(metadata.get(DISK_NAME_GEN_KEY, 0)) + 1
+        metadata[DISK_NAME_GEN_KEY] = str(gen)
+        return "{0}-{1}".format(self.instance, gen)
+
+    def createBootDisk(self, disk_name):
+        if self.useSSD:
+            diskType = "pd-ssd"
+        else:
+            diskType = "pd-standard"
+
+        return self._gce.createDisk(
+            image=self.image,
+            name=disk_name,
+            type=diskType)
+
+    @defer.inlineCallbacks
+    def start_instance(self, build):
+        instance_state = yield self._gce.processRequest(
+            self._gce.getInstanceState())
+
+        fingerprint, metadata = self.getMetadataFromState(instance_state)
+        updated_metadata = self.updateMetadata(build, metadata)
+
+        instance_stop = None
+        if instance_state['status'] not in ('STOPPED', 'TERMINATED'):
+            log.info("gce: {0} is running, requesting stop".format(
+                self.instance))
+            instance_stop = self._gce.instanceStop()
+
+        if not self.resetDisk:
+            metadata['BUILDBOT_CLEAN'] = '1'
+
+        boot_disk_create = None
+        if 'BUILDBOT_CLEAN' not in metadata:
+            log.info("gce: {0} has not been reset on stop, will do it now".format(
+                self.instance))
+            boot_disk_name = self.getNewDiskName(updated_metadata)
+            boot_disk_create = self.createBootDisk(boot_disk_name)
+
+        metadata_set = self.setMetadata(fingerprint, updated_metadata)
+
+        if instance_stop is not None:
+            log.info("gce: waiting for {0} to be stopped".format(self.instance))
```

Comment:
If we change the "is running for some reason" message to `log.warn`, this one should use `log.warn` too.